### ISSUE TITLE
Is there any way to fetch the actual values on UI and save it? [Question]

### ISSUE THREAD

**Issue by happyshows** [smartinsightsfromdata/rpivotTable#12]
Hi,

I'm planning to use this package in Shiny, but I'm wondering if there's a way to "read" the user selected filter/row/column so that user can export the setting and reload it when they visit next time?

A minimal example is appreciated, thanks.

**Comment by smartinsightsfromdata** [comment]
@happyshows let me touch upon the original library pivottable structure:
pivottable has two APIs:
- pivot() is the low level API: you need to specify everything and it is "static": you specify what needs to be reported, where, how it needs to be aggregated etc. There are no pull down, select, draggable items nothing.
- pivotUI()  (used here) is the high level interface. It is in fact a dictionary of functions that underneath call pivot() but include all the facilities mentioned above and more.
As a starting point for the R integration I decided to use pivotUI(), using the out-of-the-box functions available in this great library.
If a case could be made for integrating the lower level API I will consider, but in reality it would be very similar to use other "static" table libraries, e.g. datatables, so perhaps not worth the effort (and arguably you could build a pivot table on top of datatables using a variety of low level elements for filtering dragging, summing etc.).

In conclusion what you ask it is not available within pivotUI().  The only thing you can do is set the initial parameters, rendering the pivottable the first time with a full set of parameter already on rows, cols etc. But it is part of the analytic power of the tool to let the user play with all the parameters.

**Comment by happyshows** [comment]
@smartinsightsfromdata  thanks for the details. Let me explain why I think of this idea:
I intended to use db connection + rpivotable + shiny. This solution is very generic and will mean a lot for reporting purpose, simply because excel is not a good way to reproduce a simple analysis.

I'd say in my organization, more than 50% reporting needs from regular business users can be solved with this solution, the only missing piece for users is to save their pivot table for next time, or they will go back to excel.

If the configuration values can be captured, we can: 1) save all historic pivot tables for each user. 2) enable machine learning to build reports automatically

From my perspective, the ideal UI would have two more buttons/APIs, one for importing configuration string and one for exporting.

**Comment by smartinsightsfromdata** [comment]
@nicolaskruchten from an integration standpoint, as mentioned above, pivotUI() is the "right" integration end point for R.
On the other hand I understand @happyshows use case. The ability to get the "current" set of reporting parameters (including any filtering etc.) is interesting, maybe through a call-back?
Let me know if you would like me to submit a feature request on pivottable.

**Comment by nicolaskruchten** [comment]
there is the `onRefresh` callback already... the example is here: http://nicolas.kruchten.com/pivottable/examples/onrefresh.html

**Comment by smartinsightsfromdata** [comment]
OK @happyshows this goes on my TODO list.

**Comment by happyshows** [comment]
@smartinsightsfromdata  cool, many thanks.

**Comment by happyshows** [comment]
@smartinsightsfromdata Hi, any update on this thread? If possible, could you provide an ETA for completion? I'd like to plan whether this package will be on next phase of my project or not....

**Comment by williFiebranz** [comment]
Dear Nicolas,

can you give an example how I would use the onRefresh argument from within a shiny app in order to get the JSON that describes the current state of the pivotTable?

Thank you and good bye

**Comment by smartinsightsfromdata** [comment]
@williFiebranz   the ```onRefresh``` callback is currently not implemented in rpivottable.

**Comment by williFiebranz** [comment]
Okay, thanks for the quick reply. Is this feature planned to be implemented at some point?

**Comment by smartinsightsfromdata** [comment]
@williFiebranz to implement this feature is a bit tricky, mainly because relies on a call back.  Htmlwidgets has a minimal callback ability (I've used it for another widget), but is going to get a bit of time to get it right.
I may have more time in August, but I will accept pull requests if anybody has more time.

**Comment by williFiebranz** [comment]
That sounds great!

**Comment by dkilfoyle** [comment]
Kudos to smartinsightsfromdata, pivottable.js and ramnathv for a great tool. I also wanted to extract data from the pivottable. Here is one way:

Append these two lines of code in rpivottable.js just before the call to pivotUI:
```
      if (typeof x.params.onRefresh != "undefined") {
        x.params.onRefresh = x.params.onRefresh[0];
      }
```
Then in your R call to rpivottable add an option:
```
onRefresh=htmlwidgets::JS("function(config) { Shiny.onInputChange('myPivotData', config); }"
```
Here is an example shiny app that prints the selected columns:
```
server <- function(input, output) {
  
  output$pivotRefresh = renderText({
    x=input$myPivotData
    paste(x$cols)
  })
  
  output$mypivot = renderRpivotTable({
    rpivotTable(data=cars, onRefresh=htmlwidgets::JS("function(config) { Shiny.onInputChange('myPivotData', config); }"))
  })
}

ui <- fluidPage(
  sidebarLayout(
    sidebarPanel(),
    mainPanel(
      rpivotTableOutput("mypivot"), 
      textOutput("pivotRefresh")
    )
  )
)

shinyApp(ui = ui, server = server)
```

**Comment by smartinsightsfromdata** [comment]
@dkilfoyle this looks fantastic.  Too bad I submitted the current version just yesterday night to CRAN!

I'll post the version on github on a dev branch later today.  I cannot re-submit to CRAN till I hear from them.

**Comment by smartinsightsfromdata** [comment]
@williFiebranz @happyshows 
I've posted on stackoverflow (for the time being) an example about this functionality (a bit more complete than @dkilfoyle one, but I am still not completely satisfied about it).
The code is for the time being on the dev branch (see stackoverflow for details).

Any comment is welcome, especially on how to use it for the greater good...
 
Again, thanks to @dkilfoyle that contributed this feature.

and as always thanks to @nicolaskruchten for the library :)

**Comment by happyshows** [comment]
@smartinsightsfromdata thanks, could you post a link to the stackoverflow page?

**Comment by smartinsightsfromdata** [comment]
@happyshows  gosh!.  I forgot the most important thing. So sorry.

[Here is the link](http://stackoverflow.com/questions/33214397/download-rpivottable-ouput-in-shiny/33352302#33352302)

**Comment by smartinsightsfromdata** [comment]
Changes now in master.  I've put among the examples the shiny code.  Thanks again to @dkilfoyle

**Comment by dkilfoyle** [comment]
Welcome :-) Thanks for bringing this great library to R.

Sent from my Samsung Galaxy smartphone.-------- Original message --------From: Enzo <notifications@github.com> Date: 26/10/2015  22:35  (GMT-10:00) To: smartinsightsfromdata/rpivotTable <rpivotTable@noreply.github.com> Cc: dkilfoyle <dkilfoyle@xtra.co.nz> Subject: Re: [rpivotTable] Is there any way to fetch the actual values on UI
  and save it? [Question] (#12) 
Changes now in master.  I've put among the examples the shiny code.  Thanks again to @dkilfoyle
